design/design-doc-resource-claim-definition.md

```diff
@@ -0,0 +1,174 @@
+# Resource Claim and Binding Definition
+* Owner: Nic Cope (@negz)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+See the following issues this design doc intends to address:
+
+* https://github.com/crossplaneio/crossplane/issues/1105
+* https://github.com/crossplaneio/crossplane/issues/1106
+* https://github.com/crossplaneio/crossplane/issues/1107
+
+TODO(negz): Write this up.
+
+## Goals
+
+* Allow infrastructure operators to define claim kinds.
+* Allow a claim to be satisfied by multiple resources.
+* Surface claim to resource compatibility in Crossplane's API.
+
+TODO(negz): Write this up.
+
+## Proposal
+
+This document proposes the introduction of two new, cluster scoped, resources
+named `ResourceClaimDefinition` and `ResourceBindingDefinition`.
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClaimDefinition
+metadata:
+  name: mysqlinstance
+spec:
+  group: database.crossplane.io
+  names:
+    kind: MySQLInstance
+    listKind: MySQLInstanceList
+    plural: mysqlinstances
+    singular: mysqlinstance
+  version: v1alpha1
+  validation:
+    openAPIV3Schema:
+      description: A MySQLInstance is a portable resource claim that may be satisfied
+        by binding to a MySQL managed resource such as an AWS RDS instance or a GCP
+        CloudSQL instance.
+      properties:
+        spec:
+          properties:
+            engineVersion:
+              description: EngineVersion specifies the desired MySQL engine version,
+                e.g. 5.7.
+              enum:
+              - "5.6"
+              - "5.7"
+              type: string
+```
+
+```yaml
+---
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceBindingDefinition
```

Comment:
Could you elaborate the benefits of not embedding `ResourceBindingDefinition`s into the `ResourceClaimDefinition`? I feel like having only `ResourceClaimDefinition` type that also defines the bindings is more concise and gives the user the comfort of checking only one place to see what the possible outcomes if they were to create a `ResourceClaim`.

Reply:
I actually had them embedded at first as you suggest, but @bassam and @suskin pointed out that it's possible the claim schema author won't be the same person as the binding author. For example we might have a stack that publishes a few common claim schemas, but doesn't enable binding to them. Then later an infrastructure operator might want to add binding definitions, but only for the claims that are interesting to them.

Will definitely add this context to the document if we go down this path.

Reply:
I've given some more thought to this and I think I agree with you guys.

When an application stack developer makes decision of what infrastructure to use, they think of the claim _type_ as the medium between the app and the infrastructure. Claim is basically the abstraction layer. In this context, as long as the resource usage pattern (basically connection secret's shape) stays the same, infra operator should be able to change what you _actually_ get when you create that claim, which is way easier with separate `ResourceBindingDefinition`s.